tests/test_results.py

```diff
@@ -249,3 +250,32 @@ def test_job_return_value_unserializable(self):
 
         Result.create(job, Result.Type.SUCCESSFUL, ttl=10, return_value=1)
         self.assertEqual(Result.count(job), 2)
+
+    def test_blocking_results(self):
+        queue = Queue(connection=self.connection)
+        job = queue.enqueue(say_hello)
+
+        # Should block if there's no result.
+        timeout = 1
+        result_sync = Result.fetch_latest(job)
```

Comment:
This just crossed my thought: should we check whether the job has started or finished before waiting for result? If job has not been started, should it raise an exception?

Reply:
> If job has not been started, should it raise an exception?

My instinct is no.

Typically I just want to "fire and wait", without really caring if I'm waiting because my workers are busy with other jobs vs waiting for execution to finish. Do you have an example in mind where that difference would be meaningful?

Also I'll admit my motivation for adding this PR was to reduce the latency and performance hit of polling for results. Adding an exception for waiting for an un-started job would mean either going back to polling for a started job before blocking for results, or adding a new `Job.block_until_started()` method.


> should we check whether the job has started or finished before waiting for result?

I think regardless of the job status we'd need to query the result:

* If the job is `FINISHED`/`FAILED` we need to get the result.
* If the job is `QUEUED`or otherwise pending, by the time `Job.get_status()` returns, the job could have finished with a Result by now, so we need to check for a result.
* If the job is  `CANCELLED`/`STOPPED` I think there can still be a result from a prior retry attempt?

So yes if the job is `FINISHED` we could use non-blocking `xrevrange` rather than `xread`, which is a cheaper redis function. But to get the status requires a `hget` call which removes any performance advantage.

Reply:
I agree with your reasoning. I'll take another look soon.